### DISCUSSION TITLE
How to load TFF from Content? (Как загрузить TFF-шрифты из контента)

### DISCUSSION THREAD

**Discussion by kreghek** [rds1983/FontStashSharp#67] [Q&A] [answered]
Can you get sample loading tff fonts from Monogame Content Pipeline? For example, `Content.Load<byte[]>("Droid")`. I can't understand some things:
- Can I avoid coping of the tff to bin folder and load it directly instead using the content pipeline? I am using NoPipeline library in my project and want to use single approach to work with content.
- Which importer and processor can be used for it?

thank!

(Можете показать пример, в котором шрифты загружаются через `Content.Load<>()`. Например, так `Content.Load<byte[]>("Droid")`. Сам не могу понять:
- Можно ли избежать прямой загрузки tff через `File.ReadAllBytes`? Ведь чтобы так загрузить файл, нужно его предварительно скопировать в bin-директорию. Я сейчас использую библиотеку NoPipeline. Контент вынесен в отдельный проект. И не хотелось быть иметь отдельный подход при загрузке контента только для шрифтов.
- Думаю, всё сводится к указанию правильных импортера и процессора при обработке tff в составе контента. Но сам я не понимаю, какие использовать. И возможно, вообще ошибаюсь с этим предположением.
Спасибо!)

**Comment by Beyley** [comment]
> Can you get sample loading tff fonts from Monogame Content Pipeline? For example, `Content.Load<byte[]>("Droid")`. I can't understand some things:
> - Can I avoid coping of the tff to bin folder and load it directly instead using the content pipeline? I am using NoPipeline library in my project and want to use single approach to work with content.
> - Which importer and processor can be used for it?
> 
> thank!
> 
> (Можете показать пример, в котором шрифты загружаются через `Content.Load<>()`. Например, так `Content.Load<byte[]>("Droid")`. Сам не могу понять:
> - Можно ли избежать прямой загрузки tff через `File.ReadAllBytes`? Ведь чтобы так загрузить файл, нужно его предварительно скопировать в bin-директорию. Я сейчас использую библиотеку NoPipeline. Контент вынесен в отдельный проект. И не хотелось быть иметь отдельный подход при загрузке контента только для шрифтов.
> - Думаю, всё сводится к указанию правильных импортера и процессора при обработке tff в составе контента. Но сам я не понимаю, какие использовать. И возможно, вообще ошибаюсь с этим предположением.
> Спасибо!)

you can just use File.RealAllBytes and manually pass in the path to the file in the content folder

**Reply by kreghek** [reply]
To use `File.RealAllBytes` I must copy tff files to content directory (in the bin directory) manually. It's become more complicated then the content files in the separeted project (to share and reuse). Also I'm not sure it works in `dotnet publish` without some workarounds.

I thought I can handle tff-files in the same way as other content files like png or ogg.

**Reply by Beyley** [reply]
> To use `File.RealAllBytes` I must copy tff files to content directory (in the bin directory) manually. It's become more complicated then the content files in the separeted project (to share and reuse). Also I'm not sure it works in `dotnet publish` without some workarounds.
> 
> I thought I can handle tff-files in the same way as other content files like png or ogg.
you can tell the content pipeline to copy files

**Reply by kreghek** [reply]
> you can tell the content pipeline to copy files

So, this is acceptable solution. How to do it? Just set no importer and no processor?

**Reply by rds1983** [reply]
Another solution is to ignore the Content Pipeline. And include font files directly in the project. And set "Copy to Output Directory" to "Copy if newer". That's what FSS samples do.
Or you could include font files as resources in the project.

**Comment by kreghek** [comment, marked as answer]
The most simple solution is:

```c#

// game content extentions project

public class TtfImporter : ContentImporter<byte[]>
{
    public override byte[] Import(string filename, ContentImporterContext context)
    {
        return File.ReadAllBytes(Path.ChangeExtension(filename, ".ttf"));
    }
}

public class TtfProcessor : ContentProcessor<byte[], byte[]>
{
    public override byte[] Process(byte[] input, ContentProcessorContext context)
    {
        context.Logger.LogMessage($"Do nothing with binary.");

        return input;
    }
}

// game client project

_fontSystem = new FontStashSharp.FontSystem();
_fontSystem.AddFont(Game.Content.Load<byte[]>("Fonts/DroidSans"));
```

It works fine with NoPipeline library:
```json
{
	"root": "relative-path-to-separatd-content-project",
	"references":
	[
		// Reference paths are not affected by root.
		"relative-path-to-game-content-extensions/bin/Debug/*.dll"
	],
	"content":
	{
                "Fonts":
		{
			"path": "Fonts/*.ttf",
			"recursive": "True",
			"action": "build",
			"importer": "TtfImporter",
			"processor": "TtfProcessor"
		}
```
